writer.go

```diff
@@ -19,7 +19,25 @@ type Writer struct {
 	framer    Framer
 	formatter Formatter
 
-	conn serverConn
+	conn          serverConn
+	needReconnect bool
+}
+
+// Poll the socket to find out if/when it closes
+func (w *Writer) poll() {
+	for {
+		if w.conn == nil {
+			w.needReconnect = true
+			return
+		}
+
+		b := make([]byte, 1)
+		_, err := w.conn.read(b)
```

Comment:
The common case is going to be that socket is open, which means a read will return and the loop will continue. For instance, I added a log line to this loop, and over the course of just this one test it executes almost 50 times on my machine. (Your number will vary depending on how fast your computer is.)